File: ts-service/src/publisher.ts
import type { Connection } from './connection.js';
import type { Validator } from './validate.js';
import { EXCHANGE, routingKeyFor, type MusicEvent } from './routing.js';

const SERVICE_NAME = 'ts-service';

/**
 * Validates an event against the AsyncAPI contract, then publishes it to the
 * topic exchange using a routing key derived from its city + genre.
 */
export class Publisher {
  constructor(
    private readonly conn: Connection,
    private readonly validator: Validator,
  ) {}

  /** Returns the routing key used. Throws if invalid and force is false. */
  publish(event: MusicEvent, force = false): string {
    const errors = this.validator.errorsFor(event);
    if (errors.length > 0) {
      const detail = errors.map((e) => `  - ${e}`).join('\n');
      if (!force) {
        throw new Error(`event fails the AsyncAPI MusicEvent contract:\n${detail}`);
      }
      console.warn(`[${SERVICE_NAME}] ⚠ publishing INVALID event anyway (--force):\n${detail}`);
    }

    const routingKey = routingKeyFor(event);
    this.conn.channel.publish(EXCHANGE, routingKey, Buffer.from(JSON.stringify(event)), {
      contentType: 'application/json',
      persistent: true,
      type: 'music.event.published',
      appId: SERVICE_NAME,
    });
    console.log(`[${SERVICE_NAME}] → published ${routingKey}  (${event.title ?? '?'})`);
    return routingKey;
  }
}

File: ruby-service/lib/publisher.rb
# frozen_string_literal: true

require 'json'

# Validates an event against the AsyncAPI contract, then publishes it to the
# topic exchange using a routing key derived from its city + genre.
class Publisher
  def initialize(connection, validator, service_name: 'ruby-service')
    @connection = connection
    @validator = validator
    @service_name = service_name
  end

  # Returns the routing key the message was published with.
  # Raises ArgumentError if the event is invalid and force is false.
  def publish(event, force: false)
    errors = @validator.errors_for(event)
    unless errors.empty?
      message = "event fails the AsyncAPI MusicEvent contract:\n  - #{errors.join("\n  - ")}"
      raise ArgumentError, message unless force

      warn "[#{@service_name}] ⚠ publishing INVALID event anyway (--force):\n  - #{errors.join("\n  - ")}"
    end

    routing_key = Routing.routing_key_for(event)
    @connection.exchange.publish(
      JSON.generate(event),
      routing_key: routing_key,
      content_type: 'application/json',
      persistent: true,
      type: 'music.event.published',
      app_id: @service_name
    )
    puts "[#{@service_name}] → published #{routing_key}  (#{event['title'] || '?'})"
    routing_key
  end
end
